### PR TITLE
Improve the execution speed of the DA clusterizers for primary vertex finding

### PR DESCRIPTION
**Pull request by werdmann**
Last year the execution time of  DAClusterizerInZ_vect and DAClusterizerInZT_vect  was improved significantly by reducing the list of clusters considered for a triack to within 4 sigma in the update() cycles.  Similar arithmetic is used in the purge() method and the final track assignment at the end of vertices().  This code is executed much less frequently than update but there is still some time to gain because they are executed at low Temperature when the cluster lists are long. 
In addition the code was streamlined to avoid repeating some computations. Only changes that are not expected to change the result have been implemented. The cluster list restriction represents an approximation (cutting off a gaussinan at 4 sigma) that can lead to very small changes.
 This PR reduces the clustering time by about 30 % with negligible changes of the results.

Small additional changes that don't change the behaviour are packaged with this PR
*  new configurable parameter (uniqutrkminp) was introduced for future use. When set to 0 (=current default) it restores the original behaviour
* the 4d vertices until now have no track weights. These are now copied from the 3d fit after the 4d clustering (in plugin/PrimaryVertexProducer.cc)

clustering-time-offlinePrimaryVertices

#### PR validation:
the PR was tested with several RelVal samples from Run 3 and Phase2. Almost no differences in the results have been found.
runthematrix reports 37 36 35 26 17 4 1 1 1 tests passed, 0 0 0 0 0 0 0 0 0 failed

### CODE REVIEW

**Comment by slava77** [comment]
enable profiling

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by perrotta** [comment]
@cmsbuild please test

**Comment by perrotta** [comment]
@werdmann please expand the title to something more descriptive: at least, include "DA" on it

**Comment by werdmann** [comment]
This seems stalled. Am I missing something that I need to do now?

**Comment by perrotta** [comment]
@smuzaffar tests seems to be stuck (I don't even see the usual temporary status report): was it the eanbling of the profile who messed them up? Did they ever started?

**Comment by smuzaffar** [comment]
please test
lets retry

File: RecoVertex/PrimaryVertexProducer/src/DAClusterizerInZ_vect.cc [hunk 1]
```diff
@@ -346,8 +342,10 @@ double DAClusterizerInZ_vect::update(
       // same loop but without updating sWE
 #pragma GCC ivdep
       for (unsigned int k = kmin; k < kmax; ++k) {
-        vertices.se[k] += vertices.exp[k] * (tmp_trk_tkwt * o_trk_sum_Z);
-        auto w = vertices.rho[k] * vertices.exp[k] * (tmp_trk_tkwt * o_trk_sum_Z * o_trk_dz2);
+        //vertices.se[k] += vertices.exp[k] * (tmp_trk_tkwt * o_trk_sum_Z);
```
Comment (perrotta) [inline RecoVertex/PrimaryVertexProducer/src/DAClusterizerInZ_vect.cc:345]: These commented out lines can be removed

File: RecoVertex/PrimaryVertexProducer/src/DAClusterizerInZ_vect.cc [hunk 2]
```diff
@@ -371,17 +369,15 @@ double DAClusterizerInZ_vect::update(
 
   // loop over tracks
   for (auto itrack = 0U; itrack < nt; ++itrack) {
-    unsigned int kmin = gtracks.kmin[itrack];
-    unsigned int kmax = gtracks.kmax[itrack];
+    const unsigned int kmin = gtracks.kmin[itrack];
```
Comment (perrotta) [inline RecoVertex/PrimaryVertexProducer/src/DAClusterizerInZ_vect.cc:372]: The same `const` can be applied in the analogous assignments in DAClusterizerInZT_vect.cc

**Review by perrotta — COMMENTED** [review]
I'm even more convinced that having unified the code for DAClusterizerInZ_vect.cc  and DAClusterizerInZT_vect.cc would have enormously simplified the parallel updates of the two codes (as well as the review).
Anyhow, let have a look at the results of the test first, then we can proceed with the following, indeed minor, adjustments

**Comment by perrotta** [comment]
Errors in unit test `testAlignmentOfflineValidation` is certainly correlated:
```
 ----- Begin Fatal Exception 01-Mar-2021 12:15:05 CET-----------------------
An exception of category 'Configuration' occurred while
   [0] Constructing the EventProcessor
   [1] Constructing module: class=PrimaryVertexValidation label='PVValidation'
Exception Message:
MissingParameter: Parameter 'uniquetrkminp' not found.
----- End Fatal Exception -------------------------------------------------
```

Quite likely also the crash in `testPVPlotting,` although it is less clear to me where it comes from

**Comment by mmusich** [comment]
@perrotta 

> Quite likely also the crash in testPVPlotting, although it is less clear to me where it comes from

it comes from the fact that this test `testPVPlotting` is requiring that `testAlignmentOfflineValidation` is passed first (it consumes a file produced by the latter test)

https://github.com/cms-sw/cmssw/blob/9f1ac7769abc67682da1ec99b0b13c3b83a57d2f/Alignment/OfflineValidation/test/BuildFile.xml#L8

so it's unsurprising that if one test fails also the other does.

**Comment by werdmann** [comment]
> An exception of category 'Configuration' occurred while
>    [0] Constructing the EventProcessor
>    [1] Constructing module: class=PrimaryVertexValidation label='PVValidation'
> Exception Message:
> MissingParameter: Parameter 'uniquetrkminp' not found.
> ----- End Fatal Exception -------------------------------------------------
This is the new parameter, indeed. But why does it fail here (and not everywhere else) ? What needs to be done beside fillDescription?

**Comment by mmusich** [comment]
To solve the issue:

```
----- Begin Fatal Exception 01-Mar-2021 12:15:05 CET-----------------------
An exception of category 'Configuration' occurred while
   [0] Constructing the EventProcessor
   [1] Constructing module: class=PrimaryVertexValidation label='PVValidation'
Exception Message:
MissingParameter: Parameter 'uniquetrkminp' not found.
----- End Fatal Exception -------------------------------------------------
```

this test configuration is using a clone of `DA_vectParameters` from `from RecoVertex.PrimaryVertexProducer.TkClusParameters_cff`

https://github.com/cms-sw/cmssw/blob/9f1ac7769abc67682da1ec99b0b13c3b83a57d2f/Alignment/OfflineValidation/test/test_all_cfg.py#L288

https://github.com/cms-sw/cmssw/blob/master/Alignment/OfflineValidation/test/test_all_cfg.py#L252-L253

I suppose one has to add `uniquetrkminp` here:

https://github.com/cms-sw/cmssw/blob/394c95a7fd61cf51eaa9c30a56f3b887ff0e77a8/RecoVertex/PrimaryVertexProducer/python/TkClusParameters_cff.py#L3-L20

**Comment by mmusich** [comment]
Another option is to call `PrimaryVertexProducer::fillDescriptions(desc)` here:

https://github.com/cms-sw/cmssw/blob/master/Alignment/OfflineValidation/plugins/PrimaryVertexValidation.cc#L3728-L3760 

instead of re-implementing the same `fillDescriptions` method there.
I can take care of that in a follow-up PR.

**Comment by perrotta** [comment]
please test

**Comment by mmusich** [comment]
@werdmann 
the error in

> 
> Unit Tests
> 
> I found errors in the following unit tests:
> 
> ---> test testAlignmentOfflineValidation had ERRORS
> ---> test testPVPlotting had ERRORS
> 

changed to
```
----- Begin Fatal Exception 02-Mar-2021 07:57:03 CET-----------------------
An exception of category 'Configuration' occurred while
   [0] Constructing the EventProcessor
   [1] Validating configuration of module: class=PrimaryVertexValidation label='PVValidation'
Exception Message:
Illegal parameter found in configuration.  The parameter is named:
 'uniquetrkminp'
You could be trying to use a parameter name that is not
allowed for this plugin or it could be misspelled.
----- End Fatal Exception -------------------------------------------------
```
so we need indeed to add the new parameter also in the `PrimaryVertexValidation` `fillDescriptions` method here:

https://github.com/cms-sw/cmssw/blob/9f197f8c7d41adde9dcf279c49203f22decb7e2f/Alignment/OfflineValidation/plugins/PrimaryVertexValidation.cc#L3748

I've tried to do as suggested in https://github.com/cms-sw/cmssw/pull/32994#issuecomment-788086014 to avoid you the pain to always update both `fillDescrptions` methods here: https://github.com/cms-sw/cmssw/compare/master...mmusich:usePrimaryVertexProducerFillDescriptions but this doesn't compile (or better it does not link), I think because `PrimaryVertexProducer` is a plugin.
Perhaps @perrotta has a good suggestion to propagate the `fillDescriptions` method of `PrimaryVertexProducer`  to `PrimaryVertexValidation`  ?

**Comment by werdmann** [comment]
Since this parameter is intended for the next round of optimization and not vital for this PR I could also just remove it now and re-introduce it later when this problem is resolved.

> @werdmann
> the error in
> 
> > Unit Tests
> > I found errors in the following unit tests:
> > ---> test testAlignmentOfflineValidation had ERRORS
> > ---> test testPVPlotting had ERRORS
> 
> changed to
> 
> ```
> ----- Begin Fatal Exception 02-Mar-2021 07:57:03 CET-----------------------
> An exception of category 'Configuration' occurred while
>    [0] Constructing the EventProcessor
>    [1] Validating configuration of module: class=PrimaryVertexValidation label='PVValidation'
> Exception Message:
> Illegal parameter found in configuration.  The parameter is named:
>  'uniquetrkminp'
> You could be trying to use a parameter name that is not
> allowed for this plugin or it could be misspelled.
> ----- End Fatal Exception -------------------------------------------------
> ```
> 
> so we need indeed to add the new parameter also in the `PrimaryVertexValidation` `fillDescriptions` method here:
> 
> https://github.com/cms-sw/cmssw/blob/9f197f8c7d41adde9dcf279c49203f22decb7e2f/Alignment/OfflineValidation/plugins/PrimaryVertexValidation.cc#L3748
> 
> I've tried to do as suggested in [#32994 (comment)](https://github.com/cms-sw/cmssw/pull/32994#issuecomment-788086014) to avoid you the pain to always update both `fillDescrptions` methods here: [master...mmusich:usePrimaryVertexProducerFillDescriptions](https://github.com/cms-sw/cmssw/compare/master...mmusich:usePrimaryVertexProducerFillDescriptions) but this doesn't compile (or better it does not link), I think because `PrimaryVertexProducer` is a plugin.
> Perhaps @perrotta has a good suggestion to propagate the `fillDescriptions` method of `PrimaryVertexProducer` to `PrimaryVertexValidation` ?

**Comment by mmusich** [comment]
I think it's better to address it here, and then I can eventually clean it up later (no guarantee it will work in either case)

**Comment by perrotta** [comment]
One possible solution could be to define a `fillPSetDescription(...)` inside one of the `DAClusterizer`'s, and call it from within PrimaryVertexProducer, PrimaryVertexValidation, and whoever else needs it

**Comment by perrotta** [comment]
Intermediate observation, while waiting for the solution of the issues with the configurations in the validations:

- Igprof CPU profiles run in the automatic tests of this PR (https://cmssdt.cern.ch/SDT/cgi-bin/igprof-navigator/CMSSW_11_3_X_2021-02-28-2300/slc7_amd64_gcc900/profiling/23434.21/PR-6312fe/13156/igprofCPU_step3) when compared with the same profiling in the IB (https://cmssdt.cern.ch/SDT/cgi-bin/igprof-navigator/CMSSW_11_3_X_2021-02-28-2300/slc7_amd64_gcc900/profiling/23434.21/igprofCPU_step3) confirm that the DAClusterizer's move back in the rank order.
For example, the DAClusterizerInZT_vect:clusterize() method moves from being 79th in the IB to becoming 129th with this PR: 
DAClusterizerInZ_vect:clusterize() goes from 181th to 275th.
- Memory profiles (privately run) do not show significant differences in the MEM_LIVE consumed by the PrimaryVertexProducer between the two cases

**Comment by perrotta** [comment]
enable none

**Comment by perrotta** [comment]
@cmsbuild please test

**Comment by yuanchao** [comment]
+1
* Though there are deprecated edm::eventsetup::EventSetupRecord::get<edm::ESHandle<xxx> > function calls in PrimaryVertexProducer.cc, they are not part of this PR. (any update foresee?)
* matrix tests and unit tests all passed

**Comment by werdmann** [comment]
@yuanchao
>     * Though there are deprecated edm::eventsetup::EventSetupRecord::get<edm::ESHandle > function calls in PrimaryVertexProducer.cc, they are not part of this PR. (any update foresee?)

This is at least not int the original workplan and I am clearly not aware of such an issue. Can you be a bit more specific about the changes that are required/recommend ?

**Comment by perrotta** [comment]
+1
- Execution speed had been improved, as planned, with apparently no negative effect on memory, as well as on the physics content
- Tiny numerical differences in outputs are visible in the test outputs, as expected

**Comment by silviodonato** [comment]
+1

**Comment by silviodonato** [comment]
@werdmann all modules should be migrated away from `EventSetupRecord::get`, that is deprecated, before 12_1_0.
You can find more info in https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideHowToGetDataFromES#Getting_data_from_EventSetup_wit .
This was not part of your PR, but it should be done sooner or later by somebody.

**Pull request merged**